lucene/core/src/test/org/apache/lucene/index/TestFieldInfos.java

```diff
@@ -187,4 +187,23 @@ public void testMergedFieldInfos_singleLeaf() throws IOException {
     writer.close();
     dir.close();
   }
+
+  public void testFieldNumbersAutoIncrement() {
+    FieldInfos.FieldNumbers fieldNumbers = new FieldInfos.FieldNumbers("softDeletes");
+    for (int i = 0; i < 10; i++) {
+      fieldNumbers.addOrGet("field" + i, -1, IndexOptions.NONE, DocValuesType.NONE,
+          0, 0, 0, 0,
+          VectorValues.SearchStrategy.NONE, false);
+    }
+    int idx = fieldNumbers.addOrGet("EleventhField", -1, IndexOptions.NONE, DocValuesType.NONE,
+        0, 0, 0, 0,
+        VectorValues.SearchStrategy.NONE, false);
+    assertEquals("Field numbers 0 through 9 were allocated", 10, idx);
+
+    fieldNumbers.clear();
```

Comment:
My only problem with unit tests like this is that it doesn't test what we _really_ want to know -- that when IW.deleteAll() is called (a user level thing, fieldNumbers.clear() is not), that the field numbers get reset.

Reply:
Uhm, but this is the purpose of a unit-test?

It should test the lowest "unit" of the software, in this case `FieldInfos` and its `clear` method.

The fact that `IW.deleteAll()` is the "user space manifestation" of this particular low level problem is actually unimportant, from the standpoint of this unit test.  The unit test should only test the smallest / shrink-wrapped problem, which is indeed that `FieldInfos.clear()` failed to do its job correctly.

We could consider adding a unit test that "IW.deleteAll does not leak memory", but I don't think that is required for committing this fix to an obvious bug  ...

Reply:
I agree that nothing more than what was done is _required_.  My feedback reflects my low opinion of the value of some unit tests like this one.  I've seen low level unit tests before that seem nice at a glance yet the component still doesn't work as intended due to other reasons.  From the conversation over the course of today I see here... it seems my point has some credence ;-)